crypto/txt_db/txt_db.c

```diff
@@ -160,6 +160,10 @@ int TXT_DB_create_index(TXT_DB *db, int field, int (*qual) (OPENSSL_STRING *),
         db->error = DB_ERROR_MALLOC;
         return 0;
     }
+    OPENSSL_LH_set_thunks((OPENSSL_LHASH *)idx, lh_OPENSSL_STRING_hash_thunk,
+                           lh_OPENSSL_STRING_comp_thunk,
+                           lh_OPENSSL_STRING_doall_thunk,
+                           lh_OPENSSL_STRING_doall_arg_thunk);
```

Comment:
It seems to me that we should never call OPENSSL_LH_new() directly.

Reply:
We don't but this is a public API so we cannot prevent third parties to do so.

Reply:
yes, no one should ever call OPENSSL_LH_new directly, but we can't prevent it based on how the LHASH macros are implemented.  In a future release we should create an oss_lhash api that is private and deprecate this, but for now, this is what we have to work with.

Reply:
Just above this is an OPENSSL_LH_new() call, which is why OPENSSL_LH_set_thunks() is called here manually.

Reply:
we do need thunks here, as far as I can see.  Some notes:
1) TXT_DB_create_index is a public function, which accepts a hash function pointer and a cmp function pointer
2) TXT_DB_create_index is called from index_index in apps.c, which passes the same function (index_serial), as both the hash and cmp function, cast to the appropriate types with LHASH_[HASH|COMP]_FN macros
3) index_serial_[hash|cmp] are typed as:
```
static unsigned long index_serial_hash(const OPENSSL_CSTRING *a)
static int index_serial_cmp(const OPENSSL_CSTRING *a, OPENSSL_CTRING *b)
```

As such, without the thunking functions present, the lhash code will make callbacks to index_serial using the generic function pointer types:
```
unsigned long hash(cons void *)
int comp(const void *a, const void *b)
```
Because the typing of the generic function pointers (using const void *) doesn't match the passed function definitions (const OPENSSL_CSTRING *), a ubsan error will be raised at run time.

We could eliminate the thunk usage here (and in fact all thunk usages) if we altered the cmp and hash function definitions to accept void pointers directly, but doing so eliminates the type safety that the LHASH api attempts to enforce, which is what the thunking is meant to avoid, and I think breaks abi compatibility at build time (as external users of the LHASH macros would get new build warnings when the registered function signatures started failing to match

Reply:
No, please note that the apps.c code does not pass index_serial_hash/cmp functions directly but a thunk function created by the IMPLEMENT_LHASH_HASH_FN() and IMPLEMENT_LHASH_COMP_FN() macros. Please look at the LHASH_[HASH|COMP]_FN macros.

Reply:
oh, goodness, you're right.  I had just assumed that IMPLEMENT_LHASH_HASH_FN prototyped the function it created using the type passed in the macro.  With that, we dont need to set the thunk functions.  I'll remove it.